### ISSUE TITLE
Ability to get root index or manifest of an image from v1/layout

### ISSUE THREAD

**Issue by deitch** [google/go-containerregistry#814]
When I want to get the root ImageIndex or Image from a remote registry, I can call `remote.Get`. At that point, with an `Image` or `ImageIndex`, I can do whatever I need.

If the image already is downloaded to a path using v1/layout, I have to do some processing to get the same root, based on the `index.json`.

I would like an ability to get the root by a matcher, so I can do `Path.Get(someArg)` and get back either an `Image` or an `ImageIndex`, and then follow the rest of the `Path` flow just like with `remote`.

However, unlike `remote.Get()`, the possible inputs are more than just a tag->hash. It could be anything in a descriptor, since the entry in `index.json` is a list of manifests, each of which is a full `Descriptor`.

The simplest I could come up with was:

```go
type Matcher func(desc Descriptor) bool
func (p Path) FindIndex(m Matcher) (ImageIndex, error) { }
func (p Path) FindImage(m Matcher) (Image, error) { }
```

The caller provides a function that either matches or does not match the descriptor, and returns a bool. The `FindIndex`/`FindImage` then can find the right blob, process it, and return the `ImageIndex`/`Image`/.

A slightly simpler interface - one that more closely hews to `remote.Get()` and may allow code reuse, would be:

```go
type Matcher func(desc Descriptor) bool
func (p Path) Find(m Matcher) (*v1.Descriptor, error) { }
```

Happy to submit a PR once design is agreed.

As discussed offline with @jonjohnsonjr

**Comment by deitch** [comment]
Actually, the more that I think about it, I think this is asking for _two_ parts:

1. A method on `ImageIndex` to filter manifests, as in `ImageIndex.FindDescriptor(Matcher) (v1.Descriptor, error)`. I can do this now by calling `ii := ImageIndex.IndexManifest()` and then looping through `ii.Manifests`, but this would be far simpler.
1. A method on `Path` to leverage that and then return the actual `ImageIndex` or `Manifest` by looking it up after filtering.

Thoughts?

**Comment by jonjohnsonjr** [comment]
Just going to dump some thoughts as partial answers so that you're not left in the dark, but this probably deserves a rough design doc type discussion...

> 1. A method on `ImageIndex`

This seems generic enough that we wouldn't need to add this to the interface. We could just have a convenience function that does something like this.

```go
func FindDescriptor(v1.ImageIndex, Matcher) (v1.Descriptor, error)
```

My problem with this is that it's a pretty specific thing. I don't _really_ mind having a bunch of these convenience methods, but it would be great if we came up with something that was more applicable to other use cases. For example, this returns a single descriptor. What if there are multiple descriptors that match your Matcher? Do we just return the first one? Do we only look at direct children of this ImageIndex, or do we recurse?

As a strawman, I'm thinking of something like [`filepath.Walk`](https://golang.org/pkg/path/filepath/#Walk):

```go
func Walk(v1.ImageIndex, func(v1.Descriptor, error)) error
```

Perhaps we'd add something to indicate the the parent node if it recurses, with root being `nil`:

```go
func Walk(v1.ImageIndex, func(v1.Descriptor, parent *v1.Descriptor, error)) error
```

And given that various kinds of traversals may be more desirable than others, some options perhaps:

```go
// Options (we can bikeshed names):
// WithDepthFirst
// WithBreadthFirst
func Walk(v1.ImageIndex, func(v1.Descriptor, parent *v1.Descriptor, error), opt ...Option) error
```

Now, this is more cumbersome to use if you only actually care about the first matching result, and I'm happy to have two functions here, but it would be great to consider both surfaces so we can be self-consistent, e.g.:

```go
func Filter(v1.ImageIndex, Matcher, opt ...Option) ([]v1.Descriptor, error)
func Find(v1.ImageIndex, Matcher, opt ...Option) (v1.Descriptor, error)
```

Both of which could respect the traversal order.

Another point worth bringing up is our unfortunate type system. I'd like it if this could work on both `v1.Image` and `v1.ImageIndex`, so maybe we could do some stupid type checking stuff here.

I would also love to consult some mature graph traversal APIs, especially in golang but really any API would be interesting.

> I can do this now by calling ii := ImageIndex.IndexManifest() and then looping through ii.Manifests, but this would be far simpler.

It's likely that underlying implementations could more efficiently implement `Walk` than a generic implementation that does just this. We could do something like check if the `v1.ImageIndex` implements a `Walk` function and call directly into that, but threading options around might complicate this, though we could centralize these options in a single, small package.

> 2. A method on Path to leverage that and then return the actual ImageIndex or Manifest by looking it up after filtering.

Is this not already satisfied by [`v1.ImageIndex.ImageIndex(v1.Hash)` and `v1.ImageIndex.Image(v1.Hash)`](https://pkg.go.dev/github.com/google/go-containerregistry@v0.1.4/pkg/v1#ImageIndex)? They don't take a Descriptor, but presumably they would return the same result if you just pass in Descriptor.Digest.

Another unfortunate thing here is that `v1.ImageIndex` has no `Blob(Hash)` method. `layout.Path` does, so I'd like to add this to the `v1.ImageIndex` interface to make stuff like this more useful.

**Comment by jonjohnsonjr** [comment]
And just to be clear, I don't plan to block this until we find the perfect API, but I would like to figure out where this code lives first. If it's generic, we should probably put it in a generic package. I'm _okay_ with just adding a method to `layout.Path` for now, but would like to have a plan for extracting it.

**Comment by deitch** [comment]
> This seems generic enough that we wouldn't need to add this to the interface. We could just have a convenience function that does something like this.

Indeed, and I already have something like that written up. Happy to submit it, if you want.

> What if there are multiple descriptors that match your Matcher? Do we just return the first one? 

Actually, my convenience function returns all of them, and then whatever wraps it decides if it wants all or just the first. 

> Do we only look at direct children of this ImageIndex, or do we recurse?

I would do only the direct children. I don't mind a walk function, like you proposed, but I think that is much more complex, and so, if you want to walk the entire tree, repeat it. That may eventually lead to a more generic walk function, but I would start simple.

The other point is that sometimes you only _want_ the direct children. The `index.json` is a case in point. So I would want to be able to restrict it to a direct-children-only walk.

> generic package

I agree, although since this stuff is all around v1 structs/interfaces, should it not just be in v1?

> Another point worth bringing up is our unfortunate type system. I'd like it if this could work on both v1.Image and v1.ImageIndex, so maybe we could do some stupid type checking stuff here.

Yeah, I wish we could. Unfortunately, the "big sibling" of this (the type system of ocispec) has the same thing.

> Is this not already satisfied by v1.ImageIndex.ImageIndex(v1.Hash) and v1.ImageIndex.Image(v1.Hash)? They don't take a Descriptor, but presumably they would return the same result if you just pass in Descriptor.Digest.

That, actually, is precisely what my initial implementation does. Forget the naming for a second, but it basically looks like this:

```go
func imageFromRoot(p layout.Path, matcher ManifestMatcher) (v1.Image, error) {
	manifests, err := manifestsFromLayout(p, matcher)
	if err != nil || len(manifests) < 1 {
		return nil, err
	}
	rootIndex, err := p.ImageIndex()
	// of there is no root index, we are broken
	if err != nil {
		return nil, err
	}
	return rootIndex.Image(manifests[0].Digest)
}

func indexFromRoot(p layout.Path, matcher ManifestMatcher) (v1.ImageIndex, error) {
	manifests, err := manifestsFromLayout(p, matcher)
	if err != nil || len(manifests) < 1 {
		return nil, err
	}
	rootIndex, err := p.ImageIndex()
	// of there is no root index, we are broken
	if err != nil {
		return nil, err
	}
	return rootIndex.ImageIndex(manifests[0].Digest)
}
```

**Comment by jonjohnsonjr** [comment]
x-ref: https://github.com/google/go-containerregistry/pull/797

**Comment by jonjohnsonjr** [comment]
> /remove-lifecycle stale

Sorry, I think this bot is confused, but I can do this manually.
